src/shogun/kernel/CombinedKernel.cpp

```diff
@@ -61,54 +61,37 @@ void CCombinedKernel::init_subkernel_weights()
 	eigen_wt = eigen_wt.array().exp();
 	set_subkernel_weights(wt);
 }
-
-bool CCombinedKernel::init(CFeatures* l, CFeatures* r)
+bool CCombinedKernel::init_with_extracted_subsets(
+    CFeatures* l, CFeatures* r, SGVector<index_t> lhs_subset,
+    SGVector<index_t> rhs_subset)
 {
-	if(enable_subkernel_weight_opt && !weight_update)
-	{
-		init_subkernel_weights();
-	}
-
-	/* if the specified features are not combined features, but a single other
-	 * feature type, assume that the caller wants to use all kernels on these */
-	if (l && r && l->get_feature_class()==r->get_feature_class() &&
-			l->get_feature_type()==r->get_feature_type() &&
-			l->get_feature_class()!= C_COMBINED)
-	{
-		SG_DEBUG("Initialising combined kernel's combined features with the "
-				"same instance from parameters\n");
-		/* construct combined features with each element being the parameter */
-		CCombinedFeatures* combined_l=new CCombinedFeatures();
-		CCombinedFeatures* combined_r=new CCombinedFeatures();
-		for (index_t i=0; i<get_num_subkernels(); ++i)
-		{
-			combined_l->append_feature_obj(l);
-			combined_r->append_feature_obj(r);
-		}
-
-		/* recursive call with constructed combined kernel */
-		return init(combined_l, combined_r);
-	}
 
-	CKernel::init(l,r);
-	REQUIRE(l->get_feature_class()==C_COMBINED, "%s::init(): LHS features are"
-			" of class %s but need to be combined features!\n",
-			get_name(), l->get_name());
-	REQUIRE(r->get_feature_class()==C_COMBINED, "%s::init(): RHS features are"
-			" of class %s but need to be combined features!\n",
-			get_name(), r->get_name());
-	ASSERT(l->get_feature_type()==F_UNKNOWN)
-	ASSERT(r->get_feature_type()==F_UNKNOWN)
-
-	CFeatures* lf=NULL;
-	CFeatures* rf=NULL;
-	CKernel* k=NULL;
-
-	bool result=true;
+	CKernel::init(l, r);
+	REQUIRE(
```

Comment:
if you require this, you could have asked for those types in the signature of the method, no?